src/Database/LMDB/Simple/Cursor.hs

```diff
@@ -0,0 +1,211 @@
+  -> Codec v            -- ^ A @'Codec'@ for the value type.
+  -> CursorM k v a      -- ^ The cursor monad to run.
+  -> Database k v       -- ^ The database to run the cursor monad in.
+  -> Transaction mode a
+runCursorAsTransaction kcod vcod cm (Db _ dbi) = Txn $ \txn ->
+  alloca $ \kptr ->
+    alloca $ \vptr ->
+      withCursor txn dbi (\c -> runCursorM cm (CursorEnv c kptr vptr kcod vcod))
+
+{-------------------------------------------------------------------------------
+  Monadic operations
+-------------------------------------------------------------------------------}
+
+-- | Retrieve a key-value pair using a cursor Get operation.
+--
+-- The different typs of cursor Get operations are listed here:
+-- http://www.lmdb.tech/doc/group__mdb.html#ga1206b2af8b95e7f6b0ef6b28708c9127
+cget :: MDB_cursor_op -> CursorM k v (Maybe (k, v))
+cget op = CursorM $ do
+  r <- ask
+  found <- lift $ mdb_cursor_get' op (cursor r) (kPtr r) (vPtr r)
+  if found then do
+    k <- lift $ peekVal (kCodec r) (kPtr r)
+    v <- lift $ peekVal (vCodec r) (vPtr r)
+    pure $ Just (k, v)
+  else
+    pure Nothing
+
+-- | Like @'cget'@, but throws away the result.
+cget_ :: MDB_cursor_op -> CursorM k v ()
+cget_ = void . cget
+
+cpokeKey :: k -> CursorM k v ()
+cpokeKey k = CursorM $ do
+  s <- ask
+  lift $ pokeVal (kCodec s) (kPtr s) k
+
+cgetFirst :: CursorM k v (Maybe (k, v))
+cgetFirst = cget MDB_FIRST
+
+cgetNext :: CursorM k v (Maybe (k, v))
+cgetNext = cget MDB_NEXT
+
+cgetSet :: k -> CursorM k v ()
+cgetSet k = cpokeKey k >> cget_ MDB_SET
+
+cgetSetKey :: k -> CursorM k v (Maybe (k, v))
+cgetSetKey k = cpokeKey k >> cget MDB_SET_KEY
+
+-- | @'cgetSetRange' k@ positions the cursor at the first key greater than or
+-- equal to @k@. Return key + data.
+cgetSetRange :: k -> CursorM k v (Maybe (k, v))
+cgetSetRange k = cpokeKey k >> cget MDB_SET_RANGE
+
+{-------------------------------------------------------------------------------
+  Cursor folds
+-------------------------------------------------------------------------------}
+
+-- | Left fold over the full database.
+forEach ::
+     MDB_cursor_op
+  -> MDB_cursor_op
+  -> (a -> k -> v -> a)
+  -> a
+  -> CursorM k v a
+forEach first next f z = do
+    go first z
+  where
+    go op acc = do
+      x <- cget op
+      case x of
+        Just (k, v) -> go next (f acc k v)
+        Nothing     -> pure acc
+
+data Bound = Exclusive | Inclusive
+
+-- | A @'FoldRange'@ is defined by the inclusive/exclusive lower bound of the
+-- range, and the number of keys to fold.
+data FoldRange k v = FoldRange {
+    lowerBound :: Maybe (k, Bound)
+  , keysToFold :: Int
+  }
+
+-- | Analogous to @'foldM'@, though it only folds keys in the given
+-- @'FoldRange'@.
+--
+-- The fold starts from the first key in the database if @lbMay == Nothing@. If
+-- @lbMay == Just (k, b)@, then the fold starts from the first key greater than
+-- or equal to @k@, though @b@ determines whether @k@ can be present in the
```

Comment:
```suggestion
-- @lbMay == Just (k, b)@, then the fold starts from the first key @>=k@, though @b@ determines whether @k@ can be present in the
```

Reply:
> though @b@ determines whether @k@ can be present in the result or not

I don't get this: `b` is not a boolean. How does it determine this?

Reply:
`b` determines the type of lower bound, which can be exclusive or inclusive. If `b == Exclusive`, then `k` can not be in the result. I'll make this explicit in the comment